core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -71,44 +89,56 @@ public LockCache(Function<? super K, ? extends ReentrantReadWriteLock> defaultLo
     mSoftLimit = (int) Math.round(SOFT_LIMIT_RATIO * maxSize);
     mCache = new ConcurrentHashMap<>(initialSize, DEFAULT_LOAD_FACTOR, concurrencyLevel);
     mIterator = mCache.entrySet().iterator();
-    mEvictLock = new ReentrantLock();
-    mLastSizeWarningTime = 0;
+    mEvictor = Executors.newSingleThreadExecutor(
+        ThreadFactoryUtils.build(EVICTOR_THREAD_NAME, true));
+    mEvictor.submit(() -> {
+      try {
+        while (!Thread.interrupted()) {
+          evictIfOverLimit();
+        }
+      } catch (InterruptedException e) {
+        // Allow thread to exit.
+      }
+    });
   }
 
   /**
-   * If the size of the cache exceeds the soft limit and no other thread is evicting entries,
-   * start evicting entries.
+   * If the size of the cache exceeds the soft limit, evicts entries with zero references until
+   * cache size decreases below the soft limit or the whole cache is scanned.
    */
-  private void evictIfOverLimit() {
-    int numToEvict = mCache.size() - mSoftLimit;
-
-    if (numToEvict <= 0) {
-      return;
-    }
-
-    if (mEvictLock.tryLock()) {
-      try {
-        // update the total number to evict while we are holding the lock
-        numToEvict = mCache.size() - mSoftLimit;
-        // This thread won the race as the evictor
-        while (numToEvict > 0) {
-          if (!mIterator.hasNext()) {
-            mIterator = mCache.entrySet().iterator();
-          }
-          Map.Entry<K, ValNode> candidateMapEntry = mIterator.next();
-          ValNode candidate = candidateMapEntry.getValue();
-          if (candidate.mIsAccessed) {
-            candidate.mIsAccessed = false;
-          } else {
-            if (candidate.mRefCount.compareAndSet(0, Integer.MIN_VALUE)) {
-              // the value object can be evicted, at the same time we make refCount minValue
-              mIterator.remove();
-              numToEvict--;
-            }
+  private void evictIfOverLimit() throws InterruptedException {
+    try (LockResource l = new LockResource(mEvictLock)) {
```

Comment:
Do we need to hold the lock the whole time while evicting? We only have 1 evictor thread, and we only need the lock to receive the signal from other threads. Can't we just use the lock to detect if evictor should run, and then release it?

Reply:
It makes no difference. We can release the lock once the signal is received, then the `tryLock` in other `get` can lock and signal the condition, but if the eviction thread is in the while loop, the signal will be a no-op anyway. So if eviction is happening, it can keep holding the lock, because even if other threads can acquire the lock, they cannot trigger the eviction anyway.